cranelift/wasm/src/state.rs

```diff
@@ -241,6 +243,59 @@ pub struct FuncTranslationState {
     // `FuncEnvironment::make_direct_func()`.
     // Stores both the function reference and the number of WebAssembly arguments
     functions: HashMap<FuncIndex, (ir::FuncRef, usize)>,
+
+    // In each (C) function, stack arrays are allocated contiguously. We want
+    // to deterministically guarantee that no two consecutive stack allocations
+    // share the same MTE tag. Therefore, we generate an initial MTE tag
+    // randomly, and, consequently, increment that value (and apply modulo of
+    // course) and use that as the tag.
+    // We store the whole tagged index, which includes the MTE tag in its upper
+    // bits, instead of just the MTE tag.
+    latest_tagged_index: Option<Value>,
+}
+
+const MTE_NON_TAG_BITS_MASK: i64 = 0xF0FF_FFFF_FFFF_FFFFu64 as i64;
+const MTE_TAG_BITS_MASK: i64 = 0xF << 56;
+const MTE_TAG_ONE: i64 = 0x1 << 56;
+
+impl FuncTranslationState {
+    /// Tag the index with an MTE tag. If an initial tag has already been
+    /// generated, increment and use that, otherwise generate a new random tag
+    /// with IRG. Returns tagged index.
+    pub fn tag_index(&mut self, index: Value, builder: &mut FunctionBuilder) -> Value {
+        let new_tagged_index = match self.latest_tagged_index.take() {
+            None => {
+                println!("called irg");
+                // Generate random tag with IRG
+                let tagged_index = builder.ins().arm64_irg(index);
+
+                // Insert value into value storage (since values need to be typed).
+                // let tagged_index = builder.func.dfg.insert_value(I64, tagged_index);
+
+                tagged_index
+            }
+            Some(tagged_index) => {
+                // Extract random tag
+                let latest_tag = builder.ins().band_imm(tagged_index, MTE_TAG_BITS_MASK);
+
+                // Increment saved, and wrap around if 16-byte overflow would occur.
```

Comment:
I think this might be a bug. If we increment 15 (i.e. 0b1111), then we would get 0b, which is the free tag. So we have to somehow ensure we don't get any of the special tags (only Free tag equal to 0b0000 for now, but we might add more in the future).

Reply:
Right. Also once #10 is ready, we need to exclude tags that are able to access the runtime memory or memory of other instances.
We might use odd tags for `segment` instructions and even tags for everything else (runtime memory, free tags, etc.).

For this, we need to make sure `irg` only generates one of the odd tags and then simply change the above to increment by two.

Reply:
Don't you think only using odd tags is somewhat of a waste of tags, since we only need two special tags for #10. Maybe it's fine because we're guaranteeing no consecutive stack allocations have the same tag, but we'd also be giving up half the possible tag values, possibly decreasing probabilistic safety in some way?

Reply:
I'm trying to think of a simpler (probably less efficient) way of excluding our special tags. Any ideas for an algorithm for that, or is splitting in even and odd really the only performant way to go? I was thinking maybe just use an if statement on the 0b0000 tag for now, but that will get harder once we add more tags (from #10) and is also an extra cond instruction with overhead.

Reply:
Yeah, I agree it's a waste, but we'll need more than two special tags at some point. Wasmtime can run more than one wasm instance. In this case, each heap in the memory pool needs one distinct tag.
I am not 100% sure if that will work though.

Adding a branch is relatively slow, IMHO.

What you could do is determine the tag at compile time. I haven't thought about the security implications of this yet.

Reply:
How do I determine the tag at compile-time? Currently, we still rely on `irg` for initial tag generation, which is a runtime instruction. All I know at compile-time is the offset on the random tag generated by `irg`.

Reply:
If by "determining the tag at compile-time" you mean generating a random 4-bit value at compile-time, and not relying on/using `irg` at all. The first downside that comes to mind is that the random value would somehow be store inside the `.cwasm` binary. Maybe that is a problem if the guest code can somehow read that `.cwasm` binary and extract the tag, though I'm not even sure if that's of any use for such malicious guest code.
Should I try implementing it with a compile-time generated tag for now, see if it works, and if it does, we can still discuss whether the other option of using `irg` with half the available tags, split into even and odd tags?